### ISSUE TITLE
🧪 Experiment Log: Insider Detection ML Results (Exp 01–11)

### ISSUE THREAD

**Issue by holden** [razrfly/volfefe#228]
Living document — updated after each experiment completes. Tracks all results from #223.

---

## Status Table

| # | Approach | Status | AUC-ROC | P@100 | p-value | Notes |
|---|----------|--------|---------|-------|---------|-------|
| 01 | Isolation Forest | ✅ Complete | **0.7657** | 0.00 | 0.0000 | Strong signal. Hard bot exclusion key. |
| 02 | LOF | ❌ Negative | 0.4800 | 0.00 | 1.0000 | Sampling bias. Wrong paradigm for global features. |
| 03 | HDBSCAN (wallet-level) | 🔜 Next | — | — | — | Clusters wallets, not trades. Different architecture. |
| 04 | GNN (ring detection) | ⏳ Pending | — | — | — | Needs wallet-market graph build |
| 05 | XGBoost (supervised) | ⏳ Pending | — | — | — | Gate: after Exp 03 |
| 06 | Autoencoder | ⏳ Pending | — | — | — | MLX/M4 Neural Engine |
| 07 | Embeddings + FAISS | ⏳ Pending | — | — | — | Market question semantics |
| 08 | Survival Analysis | ⏳ Blocked | — | — | — | Needs hours_before_resolution fully populated |
| 09 | PyOD Ensemble | ⏳ Pending | — | — | — | Combines outputs of all above |
| 10 | LLM Hypotheses | ⏳ Pending | — | — | — | Feed anomaly scores to Claude |
| 11 | Cross-Market Correlation | ⏳ Pending | — | — | — | Wallet-market correlation matrix |

**Baseline to beat**: Current z-score system (AUC ~0.71 on scored trades, 31.9% coverage)

---

## Exp 01 — Isolation Forest (FEATURES_V2) ✅

**Run date**: 2026-03-04  **Details**: #224 | **FP Audit**: #226

### Final Results (after market-maker exclusion + price_extremity fix)

| Metric | IF | Baseline (z-score) | Delta |
|--------|----|--------------------|-------|
| AUC-ROC | **0.7657** | 0.7101 | +0.056 ↑ |
| Precision@100 | 0.0000 | 0.0000 | 0 |
| Precision@1000 | 0.0100 | 0.0080 | +0.002 ↑ |
| Precision@5000 | 0.0060 | 0.0040 | +0.002 ↑ |
| Separation | 0.0500 | 0.0726 | -0.022 |
| Permutation p | **0.0000** | — | — |

⚠️ Baseline covers only **31.9%** of trades (others scored as 0.0).

### Key findings

- **Hard market-maker exclusion** (wallets >50K trades) is critical. Adding `is_market_maker` as a feature made AUC *worse* (0.77→0.67) by polluting feature geometry. Hard exclusion: 580K trades excluded, AUC 0.7657.
- **`price_extremity` was 99.8% NULL** in DB. Now derived at load time: `(price - 0.5).abs() * 2`. Fully populated.
- **Signal is real** (p=0.0000). Confirmed insider trades are globally isolated from normal trades.
- **Precision@K = 0 at K=100**: insiders are rare (0.113%) and the model puts them in top-0.5% but not top-100. Need supervised approaches to push them higher.

### Config

- `n_estimators=500`, `contamination=0.01`, `max_samples=auto`, `random_state=42`
- 1,891,892 trades after bot exclusion | 2,147 confirmed insider trades | 14 features (FEATURES_V2)

### FP Audit finding (top-200 non-insiders)

Wallets 1–4: Venezuela market contrarians — entered 4-7h AFTER price crash. IF flagged them because they share global features with insiders (new wallet, low price, small size). Not insiders.

Wallet 5: Sophisticated 476-day macro trader. Not an insider.

**0/5 wallets were real insiders**. The FP pattern: contrarian gamblers who entered after news broke look globally similar to pre-move insiders.

---

## Exp 02 — Local Outlier Factor ❌

**Run date**: 2026-03-04  **Details**: #227

### Results

| Metric | LOF | IF (Exp 01) | Delta |
|--------|-----|-------------|-------|
| AUC-ROC | 0.4800 | 0.7657 | **-0.2857 ↓** |
| Precision@100 | 0.0000 | 0.0000 | 0 |
| Separation | -0.0000 | 0.0500 | -0.05 ↓ |
| Permutation p | **1.0000** | 0.0000 | — |

Worse than random. Signal: none.

### Why it failed

**Sampling bias**: Trained on 200K of 1.89M trades with `novelty=True`. Trades in markets not represented in the 200K sample end up far from all training neighbors → max anomaly score. Every top-20 trade is an `[Unknown Market]` data artifact.

**Conceptual mismatch**: LOF measures local density. Insider trades are NOT locally sparse — they look like other trades in the same price/timing region. Their anomaly is *global* (unusual wallet age + timing + size across the full distribution), which IF captures but LOF doesn't.

### Lessons

- IF's global isolation is the right approach for these 14 features
- LOF would need to run *within a single market* (comparing trades only to their market's own distribution) to be valid — different architecture, not just hyperparameter tuning
- Sample-based LOF with `novelty=True` is broken for this use case

### Config

- `n_neighbors=20`, `algorithm=ball_tree`, `metric=euclidean`, `novelty=True`
- Train: 200K (2,147 insiders + 197,853 non-insiders) | Score: 1,891,892 total
- Fit: 53s | Score: 476s

---

### ✅ Directory structure — sound

```
experiments/
├── shared/           # constants, data_loader, evaluator, mlflow_utils, export_data
├── data/             # parquet export (gitignored)
├── 01_isolation_forest/
├── 02_lof/
├── mlruns/           # MLflow tracking (gitignored)
└── .gitignore
```

Follows the contract from #223. Each experiment has `train.py`, `evaluate.py`, `README.md`, `results/`.

### ✅ Shared infrastructure — solid

- `data_loader.py`: FEATURES_V2, hard market-maker exclusion, clean null imputation
- `evaluator.py`: `report()`, `permutation_auc()`, `precision_at_k()`, `recall_at_k()`, `separation()`, `auc_roc()`
- `mlflow_utils.py`: thin wrapper, all experiments auto-log
- `constants.py`: FEATURES_V1 (archived), FEATURES_V2 (stable), MARKET_MAKER_THRESHOLD, CONFIRMED_INSIDER_WALLETS

### 🐛 Bug fixed: `hours_before_resolution` negative values

56,503 trades (2.29%) have `hours_before_resolution < 0` — trades placed after market resolution. The old normalization clipped these to 1.0 (maximum suspicion). Fixed: negative values now treated as NULL → filled with sentinel → normalized to 0.0 (no timing signal).

**Impact**: Exp 01 and 02 were run with the bug. Both should be re-run after data_loader fix. AUC may shift slightly.

### 🟡 .gitignore — fixed

Old: only covered `01_isolation_forest/results/`. Now: `*/results/*.{pkl,parquet,png}` — covers all experiments.

### 🟡 HDBSCAN needs wallet-level aggregation

Exp 01 and 02 operate on **trades** (1.89M rows). Exp 03 (HDBSCAN) clusters **wallets** by aggregate behavior — needs a `load_wallet_profiles()` function in `data_loader.py` and a `wallet_profiles.parquet` export. This is the main setup task before Exp 03.

### 🟡 1,297 [Unknown Market] trades (0.1%)

Trades with no matching `polymarket_markets` record — `market_question` and `end_date` are NULL, so `hours_before_resolution` is NULL for all of them (already handled by sentinel). They dominated Exp 02's top-20 due to LOF sampling bias. IF handled them correctly (not in top-20). No action needed.

---

## DB Table Decision

**Not yet.** From #223: *"Don't build this until at least one approach proves useful in backtest."*

Current state:
- IF has AUC 0.7657 and p=0.0000 — signal is statistically real
- BUT: Precision@100 = 0.0 — the model doesn't yet surface individual insider trades precisely enough to be actionable in the Elixir pipeline
- LOF: failed entirely
- We need at least Exp 05 (supervised XGBoost) to know whether we can achieve Precision@K > 0.05

**Trigger condition**: Add the table when a supervised model achieves Precision@100 > 0.10 (10 of top-100 trades are confirmed insiders). XGBoost with SHAP should get there.

**Proposed schema when ready**:
```sql
CREATE TABLE polymarket_experiment_scores (
  id          bigserial PRIMARY KEY,
  trade_id    bigint REFERENCES polymarket_trades(id),
  algorithm   varchar,       -- "isolation_forest_v2", "xgboost_v1", etc.
  score       decimal(8,6),  -- normalized 0.0–1.0
  run_id      varchar,       -- MLflow run ID
  scored_at   timestamptz DEFAULT now()
);
CREATE INDEX ON polymarket_experiment_scores (trade_id, algorithm);
```

---

## Next: Exp 03 — HDBSCAN (Wallet Clustering)

**Architecture shift**: Instead of scoring individual trades, HDBSCAN clusters **wallets** by aggregate behavior. Clusters with extreme profiles = rings. Wallets assigned label=-1 (noise/outlier) = lone-wolf anomalies.

**Setup needed before running**:
1. Add `load_wallet_profiles()` to `shared/data_loader.py`
2. Wallet features: `win_rate`, `avg_size_percentile`, `avg_hours_before_resolution`, `pct_buy`, `unique_markets`, `wallet_age_days`, `trade_count` (log-normalized), `wallet_total_volume` (log)
3. 103,035 non-bot wallets — fits in RAM easily

**Gate**: Does HDBSCAN find the Venezuela ring / Iran ring without being told to look? If it surfaces known rings as clusters, it's working.

---

## Related issues

- #223 — original research plan (11 approaches)
- #224 — Exp 01 full report
- #225 — pre-flight fixes (implemented before Exp 01 rerun)
- #226 — FP audit investigation (Venezuela wallets)
- #227 — Exp 02 LOF full report
